### PR TITLE
BVT v2 initial setup

### PR DESCRIPTION
**Pull request by niranjan92**

### CODE REVIEW

File: _common/checkHealth.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    function (err, res) {
+      if (err || !res) {
+        logger.error(
+          util.format('%s has failed api check :no response or error %s',
+            who, err)
+        );
+        return callback(true);
+      }
+
+      if (res && res.status !== 'OK') {
```
Comment (himanshu0503) [inline _common/checkHealth.js]: We should check for statusCode instead of status text.

File: _common/logging/logger.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+configLevel();
+
+function configLevel() {
+  winston.clear();
+
+  winston.add(WinstonConsoleTransport, {
+    timestamp: true,
+    colorize: true,
+    level: config.logLevel
+  }
```
Comment (himanshu0503) [inline _common/logging/logger.js]: Indent.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Optional REPL history
+.node_repl_history
+
```
Comment (shrivara) [inline .gitignore]: Remove extra new lines in this file.

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Optional REPL history
```
Comment (shrivara) [inline .gitignore]: Remove extra new lines. Order alphabetically.

File: .eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    "func-names": 0,
+    "no-plusplus": 0,
+    "no-use-before-define": 0,
+    "vars-on-top": 0,
+    "no-loop-func": 0,
+    "no-underscore-dangle": 0,
+    "no-param-reassign": 0,
+    "one-var-declaration-per-line": 0,
+    "one-var": 0,
+
```
Comment (shrivara) [inline .eslintrc.js]: Extra new line.

File: .eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    // extra rules not present in airbnb
+    "max-len": ["error", 80],
+
+    // temp turned off. If writing new file these should be turned on.
```
Comment (shrivara) [inline .eslintrc.js]: Why should this be turned on?

File: .eslintrc.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    "ActErr": true,
+    "AMQP": true,
+    "angular": true,
+    "ansi_up": true,
+    "app": true,
+    "async": true,
+    "global": true,
+    "jsyaml": true,
+    "config": true,
+    "d3": true,
```
Comment (shrivara) [inline .eslintrc.js]: Lets keep this list small to start with. Add only the ones which we are required for tests.

File: .eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
+module.exports = {
+  "env": {
+    "browser": true
```
Comment (shrivara) [inline .eslintrc.js]: True?

File: _common/setupTests.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+var self = setupTests;
+var nconf = require('nconf');
+module.exports = self;
```
Comment (himanshu0503) [inline _common/setupTests.js]: This should be on the previous line.

File: _common/micro/MicroService.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+MicroService.prototype.init = function (callback) {
+  logger.verbose('Initializing', msName);
+  async.series(
+    [
+      this.checkHealth.bind(this)
+    ],
+    function (err) {
+      if (err) {
+        logger.error('Failed to init microservice');
+        return this.error(err);
```
Comment (himanshu0503) [inline _common/micro/MicroService.js]: We can remove the retry logic. In our microservices we do a process.exit(1) if the microservice init fails.

File: _common/logging/logger.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
+'use strict';
```
Comment (shrivara) [inline _common/logging/logger.js]: Copy logger from micro which was cleaned up.

File: _common/setupTests.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+global.async = require('async');
+
+// each test starts off as a new process
+// this function does necessary setup like getting necessary data
+function setupTests(params) {
+  global.msName = params.msName;
+  process.title = params.msName;
+  global.config = {};
+
+
```
Comment (himanshu0503) [inline _common/setupTests.js]: Extra line.

File: _common/logging/logger.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+configLevel();
+
+function configLevel() {
+  winston.clear();
+
+  winston.add(WinstonConsoleTransport, {
+    timestamp: true,
+    colorize: true,
+    level: config.logLevel
```
Comment (shrivara) [inline _common/logging/logger.js]: Where is `config` coming from? Lets be explicit if its global.

File: _common/micro/MicroService.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
+'use strict';
+
+var self = MicroService;
+module.exports = self;
+
+function MicroService(params, callback) {
```
Comment (shrivara) [inline _common/micro/MicroService.js]: Lets not use prototypes. Change this to a simple function or use async.

File: testRunner.js [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+var self = start;
+module.exports = self;
+
+// Entry point for all the tests
+// setup the microservice for api health checks
+// fetch system integrations for Auth using service user token
+// will run mocha test modules
+//    https://github.com/mochajs/mocha/wiki/Using-mocha-programmatically
+
+
```
Comment (himanshu0503) [inline testRunner.js]: Extra line.

File: _common/setupTests.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+var self = setupTests;
+var nconf = require('nconf');
```
Comment (shrivara) [inline _common/setupTests.js]: Do we still need `nconf`?

File: _common/setupTests.js (hunk 12 above)
Comment (shrivara) [inline _common/setupTests.js]: Extra new line.

File: package.json [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    "underscore": "^1.8.3",
+    "request": "^2.78.0",
+    "parse-links": "*",
+    "querystring": "*",
+    "oauth": "*",
+    "chai": "^3.5.0",
+    "chai-http": "^3.0.0",
+    "util": "0.10.3",
+    "winston": "2.2.0",
+    "braintree": "1.40.0"
```
Comment (shrivara) [inline package.json]: Are we going to use this?

File: package.json [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+  },
+  "jshintConfig": {
```
Comment (shrivara) [inline package.json]: Lets use only one of es/jshint. If you want this to use es, please add the steps for people to set it up in the README. Also, add a `shippable.yml` that runs the linting.

File: testRunner.js [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+var checkHealth = require('./_common/checkHealth.js');
+var MS = require('./_common/micro/MicroService.js');
+var setupTests = require('./_common/setupTests.js');
+
+var Mocha = require('mocha'),
```
Comment (shrivara) [inline testRunner.js]: Each one should have a `var`.

File: tests/core_account_tests.js [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+var nconf = require('nconf');
+var request = require('request');
+var util = require('util');
+var setupTests = require('../_common/setupTests.js');
+var ShippableAdapter = require('../_common/shippable/Adapter.js');
+
+var tokens = {};
+var githubSysIntId = null;
+
+
```
Comment (shrivara) [inline tests/core_account_tests.js]: Extra new line.

File: tests/core_account_tests.js [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+describe('ACCT-GHC-ADM-IND.1 - Get shippable token for github login',
+  function () {
+    this.timeout(0);
```
Comment (shrivara) [inline tests/core_account_tests.js]: No magic numbers. Get this from a config.

File: tests/core_account_tests.js [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+// test login based on token and scm name
+// can be reused for any provider
+function loginUsingToken(token, scmSystemIntegrationId) {
+  if (!scmSystemIntegrationId) {
+    logger.error('failed to get system integration for scm: ',
+      scmSystemIntegrationId);
+    return;
+  }
+
+  var promise = new Promise(
```
Comment (shrivara) [inline tests/core_account_tests.js]: Are we going to use promises everywhere? This might just be easier with callbacks.

File: package.json (hunk 17 above)
Comment (niranjan92) [inline package.json]: later when we test for account cards.

Comment (niranjan92) [inline package.json]: was taken from old bvt

File: _common/setupTests.js (hunk 16 above)
Comment (niranjan92) [inline _common/setupTests.js]: nconf will be needed to persist state. will be needed when I start doing teardown.

File: tests/core_account_tests.js (hunk 22 above)
Comment (niranjan92) [inline tests/core_account_tests.js]: I used promises as the function remains generic and can be used anywhere.

else for callback, then it is limited to async series.

File: _common/logging/logger.js (hunk 11 above)
Comment (niranjan92) [inline _common/logging/logger.js]: done

File: package.json [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+  "name": "bvt-v2",
+  "version": "0.0.1",
+  "description": "Shippable Build Acceptance Tests",
+  "main": "app.js",
+  "scripts": {
+    "test": "node testRunner.js"
+  },
+  "author": "Avi Cavale <avi@shippable.com> (http://shippable.com/)",
+  "dependencies": {
+    "async": "^2.1.2",
```
Comment (rageshkrishna) [inline package.json]: Please lock down all versions. No `~`, `^` or `*`.

File: _common/logging/logger.js [hunk 24]
```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+var self = logger;
+module.exports = self;
+
+var winston = require('winston');
+
+function logger() {
+  var logger = winston;
+  // Once systemConfigs are obtained from the API, this will be overwritten.
```
Comment (shrivara) [inline _common/logging/logger.js]: This comment is not relevant here.

File: _common/setupTests.js [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+
+// each test starts off as a new process
+// this function does necessary setup like getting necessary data
+function setupTests(params) {
+  global.msName = params.msName;
+  process.title = params.msName;
+
+  // setup logger
+  global.config = {};
+  global.config.logLevel = 'verbose';
```
Comment (shrivara) [inline _common/setupTests.js]: This should come from a config file.

File: _common/setupTests.js [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+  global.logger = Logger();
```
Comment (shrivara) [inline _common/setupTests.js]: This should use the log level set in the previous step.

File: _common/logging/logger.js [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+
+  logger.add(winston.transports.Console,
+    {
+      timestamp: true,
+      colorize: true,
+      level: defaultLogLevel
+    }
+  );
+
+  logger.add(winston.transports.File,
```
Comment (shrivara) [inline _common/logging/logger.js]: Lets get rid of all the file loggers. We'll add it later if its really needed.

File: testRunner.js [hunk 28]
```diff
@@ -0,0 +1,184 @@
+'use strict';
```
Comment (shrivara) [inline testRunner.js]: Discussed with @niranjan92 that we'll add this later on when we have multiples tests in place.

File: .eslintrc.js [hunk 29]
```diff
@@ -0,0 +1,57 @@
+module.exports = {
```
Comment (shrivara) [inline .eslintrc.js]: Lets add a `shippable.yml` later to run linting on this later.

Comment (niranjan92) [inline .eslintrc.js]: will add a TODO note.

File: tests/core_account_tests.js [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+      }
+    );
+
+    it('1. Login should generate API token',
+      function (done) {
+        var json = {
+          accessToken: global.githubOwnerAccessToken
+        };
+        global.pubAdapter.postAuth(githubSysIntId, json,
+        function (err, body, res) {
```
Comment (rageshkrishna) [inline tests/core_account_tests.js]: Indent required

**Pull request merged**